Clarify docs for `Domain.recommended_domain_count`

dra27 opened the pull request:
The documentation for `Domain.recommended_domain_count` refers to domains being spawned, but this is potentially unclear in two ways:
1. If it did refer to spawns, then the value returned should be 1 less, to account for the initial domain
2. It's not explicit that if a domain has terminated you can spawn another one

I've reworded it to refer to simultaneously running domains instead (which is what it's really about - you don't want domains fighting over CPU threads). I used "running" rather than "executing" because a domain could be blocked on I/O (and so possibly be interpreted as not "executing" because it's "blocked") but still be engaged in runtime activities (marking/sweeping, STW, etc.).

I've also explicitly document the fact that the value is guaranteed to be at least 1, to be clear that that property can be relied on (for example if using it as a default value - cf. ocaml-multicore/domainslib#91):
https://github.com/ocaml/ocaml/blob/28dfeb0f61e96ab7bf1aef77e99d65ae68863416/runtime/domain.c#L1810-L1812

gasche commented:
I tried to use this value with `Domainslib.setup_pool` recently and ended up with the same questions; timely PR!

Your proposal:

```ocaml
val recommended_domain_count : unit -> int
(** The recommended maximum number of domains which should be running
    simultaneously (including the initial domain).
    The value returned is at least [1]. *)
```

I'm not sure about "(including the initial domain)"; I considered "(including already-running domains)" instead which I think it both more general and actually clearer.

Is there an easy way for OCaml programs to know the number of domains currently running? If not, should we consider adding one?

dra27 committed:
Clarify docs for Domain.recommended_domain_count

More clearly state that this is about the total number of domains
running at any given time.

Also explicitly document the fact that the value is guaranteed to be at
least 1.

dra27 commented:
Yes, that is nicer! For personal preference, I went with "x already y" rather than "already-x y", if that's OK? 🙂

> Is there an easy way for OCaml programs to know the number of domains currently running?

I don't think we expose one at the moment, no - although the information can obviously easily be returned.

> If not, should we consider adding one?

At the moment it seems surprising to me that a program which really cared about this would not be in control of working this out. If libraries need to spawn domains, they're such a scarce resource that they really should be exposing the _number_ of domains to be spawned via their API (or receiving a task pool, etc., etc.) rather than working out a number. That said, the running domain count is something which perhaps should/could be being returned in things in something like `Gc.stat`?

gasche reviewed (approved):
I think the current diff is a useful clarification. Approved.

gasche commented:
I'm curious to ask. Are domains such a scarce resource? What would go wrong if we had `2 * Domain.recommended_domain_count ()` domains, or even `3 * _`? Did people actually measure performance cliffs when several threads content for one core, or are we talking the "one per core is the right default" advice too literally? (In practice of course my computer always runs other software at the same so there are many other threads competing for the cores...)

dra27 commented:
They are scarce because of `Max_domains` (especially on 32-bit, although that's obviously not _that_ important!), not strictly because of the 1-domain-per-core "rule". My understanding is that the performance also tanks on parallel workloads once the cores are over-committed, but I don't have references for that!

xavierleroy commented:
> I'm curious to ask. Are domains such a scarce resource? 

That's not the main reason to not create more domains than available cores:

> What would go wrong if we had 2 * Domain.recommended_domain_count () domains, or even 3 * _? 

The stop-the-world GC could get a lot slower.  All the domains that are not currently running on a processor core have to be rescheduled to participate in the stop-the-world barrier, and this can take a long time if there are many such domains, esp. if the OS scheduler is tuned for throughput instead of latency.

We've seen that with the test suite running on 2-core VMs: most tests that spawn multiple domains time out frequently, to the point they had to be turned off.

gasche commented:
I understand this now (but I would be interested in measures to quantify). It makes the programming model substantially more delicate than what I thought it would be: if I understand correctly, we need a domain-spawning style that is careful not to exceed the number of cores, which is fairly different from my mental model of "they are somewhat-heavy resources so don't create too much of them please". It also means that I have absolutely no idea of the performance profile of multi-domain application on contended systems (such as: my laptop, which always runs OCaml programs in parallel with everything else).

(Thinking out loud: are there small technical changes to the runtime that could avoid this contention-cost effect? For example, could we systematically rely on the backup thread to perform the GC work of a domain, and make those backup threads high-priority / ask the system to schedule them very frequently?
How do other stop-the-world runtimes deal with contention / over-committing?)

gasche commented:
Note: the discussion tangent here is similar to the one in a domainslib thread:
  https://github.com/ocaml-multicore/domainslib/issues/77#issuecomment-1286838727

xavierleroy commented:
>  I understand correctly, we need a domain-spawning style that is careful not to exceed the number of cores

Exactly.  The name `Domain.recommended_domain_count` is pretty explicit about this already...

> which is fairly different from my mental model of "they are somewhat-heavy resources so don't create too much of them please".

Time to adjust your mental model.

> contended systems (such as: my laptop, which always runs OCaml programs in parallel with everything else).

Desktops / laptops have lots of processes, but most of them are idle most of the time (fortunately).  Plus, nowadays they have a fair number of cores, which mitigates scheduling issues to some extent.  I'm more afraid of virtual machines in the cloud, where there's some serious contention with the other VMs running on the same host, and multiple levels of scheduling.

> For example, could we systematically rely on the backup thread to perform the GC work of a domain, and make those backup threads high-priority / ask the system to schedule them very frequently?

You need superuser privileges to increase the priority of a thread or process above the default priority.  Plus, unless you use POSIX realtime scheduling, the priority is just a hint that the scheduler gladly ignores.

gasche commented:
I wrote [a small benchmark](https://gitlab.com/gasche/idle-domains-bench/-/blob/trunk/bench.ml) to measure the effect of having more domains than hardware threads on my machine, depending on whether the extra domains are compute-intensive or "mostly idle". (My machine has 4 cores and 8 hardware threads in total, `recommended_domain_count` is 8. It is running Linux.)

```
w01: [i00 2.25s (x0.93)] [i01 2.28s (x0.95)] [i02 2.37s (x0.98)] [i05 2.57s (x1.07)] [i10 2.87s (x1.19)]
w02: [i00 2.42s (x1.00)] [i01 2.50s (x1.04)] [i02 2.65s (x1.10)] [i05 3.02s (x1.25)] [i10 3.35s (x1.39)]
w03: [i00 2.93s (x1.22)] [i01 2.98s (x1.24)] [i02 3.37s (x1.40)] [i05 3.82s (x1.59)] [i10 4.30s (x1.78)]
w04: [i00 3.45s (x1.43)] [i01 3.62s (x1.50)] [i02 4.45s (x1.85)] [i05 4.68s (x1.94)] [i10 5.19s (x2.15)]
w05: [i00 4.65s (x1.93)] [i01 5.10s (x2.12)] [i02 5.18s (x2.15)] [i05 5.75s (x2.39)] [i10 7.23s (x3.00)]
w06: [i00 5.56s (x2.31)] [i01 5.84s (x2.42)] [i02 7.59s (x3.15)] [i05 7.88s (x3.27)] [i10 10.69s (x4.44)]
w07: [i00 7.83s (x3.25)] [i01 8.42s (x3.49)] [i02 8.58s (x3.56)] [i05 9.52s (x3.95)] [i10 12.23s (x5.07)]
w08: [i00 8.73s (x1.81)] [i01 10.42s (x2.16)] [i02 11.70s (x2.43)] [i05 14.36s (x2.98)] [i10 22.66s (x4.70)]
w09: [i00 28.05s (x5.82)] [i01 28.46s (x5.90)] [i02 27.45s (x5.70)] [i05 27.46s (x5.70)] [i10 33.77s (x7.01)]
w10: [i00 38.57s (x8.00)] [i01 38.08s (x7.90)] [i02 39.03s (x8.10)] [i05 38.45s (x7.98)] [i10 42.62s (x8.84)]
w11: [i00 45.77s (x9.50)] [i01 45.67s (x9.48)] [i02 44.08s (x9.15)] [i05 46.68s (x9.68)] [i10 48.77s (x10.12)]
```

Each line represents a series of run with the same number of worker domains doing the same computational work (the Knuth-Bendix benchmark of the compiler testsuite). For example the w05: line tells us about the runs with 5 worker domains. The runs on the same line correspond to different numbers of "idle domains". i00 has no idle domain, i01 has one, then 2, 5, 10.

The "x1.22" number reported corresponds to the ratio between the actual runtime of the run and the "expected ideal time" for the run. The "expected ideal time" is the same time as a 1-worker 0-idle run (the "reference run") for all tests up to 8 worker domains, and twice that time for all runs with strictly more than 8 worker domains (because at least one hardware thread needs to run two domains to completion).

The numbers show a very strong in performance when moving from 8 compute domains (exactly `recommended_domain_count`) from 9 domains (from 8.7s to 28.05s). On the other hand, adding mostly-idle domains has a noticeable but much smaller effect on performance. (For example with 5 worker domains, adding 5 idle domains gets us from 4.65s to 5.76s, despite going over recommended_domain_count.)

The code running on "idle" domains:

```ocaml
      while Atomic.get idle_run do
        List.init 10 (fun _ -> ())
        |> Sys.opaque_identity
        |> ignore;
        Unix.sleepf 1E-5
      done
```

Running the same benchmark with "really idle" idle domains (they are blocked forever waiting on a lock) shows even less cost for idle domains:

```
w01: [i00 2.26s (x1.00)] [i01 2.30s (x1.02)] [i02 2.51s (x1.12)] [i05 2.39s (x1.06)] [i10 2.61s (x1.16)]
w02: [i00 2.70s (x1.20)] [i01 2.58s (x1.15)] [i02 2.77s (x1.23)] [i05 2.79s (x1.24)] [i10 3.08s (x1.37)]
w03: [i00 3.08s (x1.37)] [i01 3.02s (x1.34)] [i02 3.16s (x1.40)] [i05 3.67s (x1.63)] [i10 4.41s (x1.96)]
w04: [i00 4.17s (x1.85)] [i01 4.10s (x1.82)] [i02 3.98s (x1.77)] [i05 4.54s (x2.02)] [i10 4.88s (x2.17)]
w05: [i00 4.74s (x2.11)] [i01 4.94s (x2.20)] [i02 4.95s (x2.20)] [i05 5.13s (x2.28)] [i10 5.54s (x2.46)]
w06: [i00 5.60s (x2.49)] [i01 6.64s (x2.95)] [i02 5.95s (x2.64)] [i05 6.08s (x2.70)] [i10 6.66s (x2.96)]
w07: [i00 6.05s (x2.69)] [i01 6.40s (x2.84)] [i02 6.82s (x3.03)] [i05 7.40s (x3.29)] [i10 12.75s (x5.67)]
w08: [i00 8.12s (x1.80)] [i01 9.31s (x2.07)] [i02 9.72s (x2.16)] [i05 11.64s (x2.59)] [i10 13.87s (x3.08)]
w09: [i00 26.88s (x5.97)] [i01 29.07s (x6.46)] [i02 27.96s (x6.21)] [i05 31.32s (x6.96)] [i10 33.25s (x7.39)]
w10: [i00 37.62s (x8.36)] [i01 37.84s (x8.41)] [i02 38.16s (x8.48)] [i05 41.08s (x9.13)] [i10 45.01s (x10.00)]
w11: [i00 45.36s (x10.08)] [i01 46.39s (x10.31)] [i02 46.96s (x10.44)] [i05 48.82s (x10.85)] [i10 52.46s (x11.66)]
```

My conclusion: it's really important not to have more compute-intensive domains than hardware threads, going over the limit will typically divide performance by 2 to 4. On linux (this probably depends on the quality of the OS scheduler), domains that are not compute-intensive have a smaller effect and less of a recommended-domain-count cliff, and completely-idle domains are very cheap, possibly thanks to good scheduling for the backup thread.
